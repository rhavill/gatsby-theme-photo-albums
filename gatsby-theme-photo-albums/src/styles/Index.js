/** @jsx jsx */
import { jsx } from 'theme-ui'
import PropTypes from 'prop-types'

const Index = ({children}) => 
  <div sx={{variant: 'indexPage'}}>
    {children}
  </div>

Index.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
}

export default Index