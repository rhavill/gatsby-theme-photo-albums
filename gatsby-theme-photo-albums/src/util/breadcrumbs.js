import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs