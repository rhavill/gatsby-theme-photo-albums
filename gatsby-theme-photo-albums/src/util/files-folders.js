const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}