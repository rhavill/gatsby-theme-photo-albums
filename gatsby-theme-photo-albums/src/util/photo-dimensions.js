export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}