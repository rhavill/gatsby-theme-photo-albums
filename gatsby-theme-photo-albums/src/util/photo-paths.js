const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension, ensureLeadingAndTrailingSlash} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => { 
    let previousUrl = null
    let previousUrlWithPage = null
    return indexedForEach(
      (file, i) => {
        const pageNumber = Math.ceil((i + 1) / photosPerPage)
        photoPaths[file.url] = {
          url: getPhotoPathWithPage(pageNumber, file.url),
          previousUrl: previousUrlWithPage,
          nextUrl: null,
          parentUrl: appendPageNumberToUrl(pageNumber, file.parentUrl),
        }
        if (previousUrl) {
          photoPaths[previousUrl].nextUrl = photoPaths[file.url].url
        }
        previousUrl = file.url
        previousUrlWithPage = photoPaths[file.url].url
      }, 
      groupedFiles[relativeDirectory]
    )
  }, keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `page-${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

const appendPageNumberToUrl = (pageNumber, url) => {
  let path = url ? url : null
  if (path && pageNumber > 1) {
    path = `${ensureLeadingAndTrailingSlash(path)}page-${pageNumber}`
  }
  return path
}

module.exports = {
  getPhotoPathsWithPages, 
  getPhotoPathWithPage,
}