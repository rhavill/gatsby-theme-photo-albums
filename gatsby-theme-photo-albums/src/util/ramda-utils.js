const addIndex = require('ramda/src/addIndex')
const complement = require('ramda/src/complement')
const compose = require('ramda/src/compose')
const forEach = require('ramda/src/forEach')
const groupBy = require('ramda/src/groupBy')
const isNil = require('ramda/src/isNil')
const map = require('ramda/src/map')
const prop = require('ramda/src/prop')

/**
 * Accepts an array of objects and returns an array containing the values of a 
 * single property from each object of the input array.
 *
 * @func
 * @sig Array -> Array
 * @param {String} property The property to extract from each object of the 
 *    input array.
 * @param {Array} objects The input array containing objects.
 * @return {Array} A simple array containing values exracted from the `property` 
 *    values of the input array.
 * @example
 *
 *      const objects = [{name: 'Hilda', age: 33}, {name: 'José', age: 45}]
 *      objectArrayToPropArray('name', objects); //=> ['Hilda', 'José']
 */
const objectArrayToPropArray = (property, objects) => map(prop(property), objects)

const mapIndexed = addIndex(map)

const indexedForEach = addIndex(forEach)

const groupByProp = property => compose(groupBy, prop)(property)

const isNotNil = complement(isNil)

module.exports = {
  objectArrayToPropArray,
  mapIndexed,
  indexedForEach,
  groupByProp,
  isNotNil,
}
