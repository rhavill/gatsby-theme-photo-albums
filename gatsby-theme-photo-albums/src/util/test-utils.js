import mergeDeepRight from 'ramda/src/mergeDeepRight'
import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'emotion-theming'

import theme from '../theme'

const create = (element, options) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{element}</ThemeProvider>, options
  )
}

const customRenderer = mergeDeepRight(renderer, {create})

const Wrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}
Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options })

// re-export everything from React Testing Library
export * from '@testing-library/react'

// override react testing library render method and custom react-test-renderer r
export { customRender as render, customRenderer as renderer }