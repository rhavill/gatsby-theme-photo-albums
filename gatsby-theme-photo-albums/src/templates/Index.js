import React from 'react'
import PropTypes from 'prop-types'
import {graphql} from 'gatsby'
import compose from 'ramda/src/compose'
import map from 'ramda/src/map'
import merge from 'ramda/src/merge'
import prop from 'ramda/src/prop'
import Layout from '../components/layout/Layout'
import Folders from '../components/folders/Folders'
import Thumbnails from '../components/thumbnails/Thumbnails'
import Pager from '../components/pagination/Pager'
import {addUrlProps} from '../util/files-folders'
import {removePathPrefix} from '../util/url-text'

const Index = ({data, location, pageContext}) => {
  const {baseUrl, currentPage, numPages} = pageContext
  const pathPrefix = data.site.pathPrefix
  const path = compose(
    removePathPrefix(pathPrefix), decodeURI
  )(location.pathname)
  const folders = compose(
    map(prop('url')),
    addUrlProps(baseUrl))(data.folders.nodes)
  const files = compose(
    map(file => merge(file, {imageData: file.childImageSharp.fixed})),
    addUrlProps(baseUrl))(data.photos.nodes)
  const folderIcon = data.folderIcon.childImageSharp.fixed

  return (
    <Layout path={path}>
      <div className="listing-page" data-testid={path} >
        <section>
          <Folders path={path} folders={folders} icon={folderIcon} />
          <Thumbnails files={files} currentPage={currentPage} />
        </section>
      </div>
      <Pager path={path} currentPage={currentPage} numPages={numPages} />
      <br/>
    </Layout>
  )
}

export const query = graphql`
  query indexQuery($skip: Int!, $limit: Int!, $relativeDirectory: String!) {
    site {
      pathPrefix
    }
    photos: allFile(
      filter: {
        sourceInstanceName: {eq: "gtpaPhotos"}, 
        relativeDirectory: {eq: $relativeDirectory}
      }, 
      sort: {fields: relativePath} 
      limit: $limit 
      skip: $skip) {
      nodes {
        relativePath
        childImageSharp {
          fixed(width: 250, height: 250, cropFocus: CENTER) {
            ...GatsbyImageSharpFixed
          }
        }      
      }
    }
    folderIcon: file(sourceInstanceName: {eq: "gtpaFolderIcon"}) {
      childImageSharp {
        fixed(width: 250, height: 250) {
          ...GatsbyImageSharpFixed
        }
      }
    }
    folders: allDirectory(
      filter: {
        sourceInstanceName: {eq: "gtpaPhotos"},
        relativeDirectory: {eq: $relativeDirectory}
      }, 
      sort: {fields: relativePath}) {
      nodes {
        relativePath
      }
    }
  }
`

Index.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired
  }).isRequired,
  data: PropTypes.shape({
    site: PropTypes.shape({
      pathPrefix: PropTypes.string.isRequired,
    }),
    folderIcon: PropTypes.shape({
      childImageSharp: PropTypes.shape({
        fixed: PropTypes.object.isRequired
      }).isRequired       
    }).isRequired,
    folders: PropTypes.shape({
      nodes: PropTypes.arrayOf(
        PropTypes.shape({
          relativePath: PropTypes.string.isRequired,
        }).isRequired   
      ).isRequired
    }).isRequired,
    photos: PropTypes.shape({
      nodes: PropTypes.arrayOf(
        PropTypes.shape({
          relativePath: PropTypes.string.isRequired,
          childImageSharp: PropTypes.shape({
            fixed: PropTypes.object.isRequired,
          })
        }).isRequired   
      ).isRequired
    }).isRequired,
  }).isRequired,
  pageContext: PropTypes.shape({
    currentPage: PropTypes.number.isRequired,
    numPages: PropTypes.number.isRequired,
    relativeDirectory: PropTypes.string.isRequired,
    baseUrl: PropTypes.string.isRequired,
  }).isRequired
}

export default Index