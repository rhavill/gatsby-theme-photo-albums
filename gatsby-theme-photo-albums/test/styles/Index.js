import React from 'react'
import {cleanup, renderer} from '../../src/util/test-utils'
import Index from '../../src/styles/Index'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Index styles', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Index><div>hello</div></Index>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
