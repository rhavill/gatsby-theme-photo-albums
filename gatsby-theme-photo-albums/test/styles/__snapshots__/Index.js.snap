// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index styles renders correctly 1`] = `
<div
  className="css-1hxgsm0-Index"
>
  <div>
    hello
  </div>
</div>
`;
